# Лабораторная работа 2-2

## 2.2.1: Reading and Writing Data

Файл [grades.csv](grades.csv) представляет собой файл с именами учащихся и буквенными оценками:

```
Alice; A
Bob; B
Robert; A
Richard; C
```

Записать данные из файла в массив. Добавьте средний балл учащегося в строку учащегося (A=4,B=3,C=2,D=1).

Подсказка: функция [strip()](https://docs.python.org/3/library/stdtypes.html#str.strip) удаляет конечные пробелы из строки.

Запишите этот файл в новый файл `grades_gpa.csv`.

Импорт библиотеки пандас

In [4]:
import pandas as pd

Загрузка файла grades.csv

In [5]:
from google.colab import files
uploaded = files.upload()

Saving grades.csv to grades.csv


In [6]:
df_grades = pd.read_csv('grades.csv')
df_grades

Alice; A
0      Bob; B
1   Robert; A
2  Richard; C

Сначала открываем исходный файл и создаем новый файл для записи, потом вычисляем средний балл, а затем записываем данные с средним баллом в новый файл

Словарь

In [10]:
gpa = {'A': 4, 'B': 3, 'C': 2, 'D': 1}

In [ ]:
with open('grades.csv', 'r') as file, open('grades_gpa.csv', 'w') as output:
    for line in file:
        name, grade = line.strip().split(';')
        gpa = {'A': 4, 'B': 3, 'C': 2, 'D': 1}.get(grade.strip(), 0)

        output.write(f'{name}; {grade.strip()}; {gpa}\n')


Открываем файл

In [12]:
# импортировать библиотеку csv
import csv

# инициализировать массив верхнего уровня
data_values = []

# откройте файл и добавьте строки в виде массивов в data_values
with open('grades.csv') as csvfile:
    filereader = csv.reader(csvfile, delimiter=';', quotechar='"')
    for row in filereader:
        # Добавление среднего балла
        row.append(gpa[row[1].strip()])
        data_values.append(row)

Создадим датафрейм и запишем в новый файл данных в новый файл

In [15]:
ocenki = pd.DataFrame(data_values, columns=['Ученик', 'Оценка буквой', 'Оценка числом'])
ocenki.to_csv('grades_gpa.csv', index=False)

Откроем датафрейм

In [16]:
ocenki

Ученик Оценка буквой  Оценка числом
0    Alice             A              4
1      Bob             B              3
2   Robert             A              4
3  Richard             C              2

## 2.2.2: Data Frames

* Рассчитайте средние сертифицированные продажи для всех альбомов.

Сначала загрузим файл hit_albums.csv

In [ ]:
from google.colab import files
uploaded = files.upload()

Saving hit_albums.csv to hit_albums.csv


Сделаем файл читаемым и выведем его

In [ ]:
import pandas as pd
hit_albums = pd.read_csv("hit_albums.csv")
hit_albums

Artist  \
0                     Michael Jackson   
1                               AC/DC   
2                          Pink Floyd   
3   Whitney Houston / Various artists   
4                           Meat Loaf   
..                                ...   
72                      Tracy Chapman   
73                              Usher   
74                    Various artists   
75                    Whitney Houston   
76                            Shakira   

                                                Album  Released  \
0                                            Thriller      1982   
1                                       Back in Black      1980   
2                           The Dark Side of the Moon      1973   
3                                       The Bodyguard      1992   
4                                     Bat Out of Hell      1977   
..                                                ...       ...   
72                                      Tracy Chapman      1988   
73                                        Confessions      2004   
74  Flashdance: Original Soundtrack from the Motio...      1983   
75                                            Whitney      1987   
76                                    Laundry Service      2001   

                          Genre  Certified sales (millions)  \
0                Pop, rock, R&B                        45.4   
1                     Hard rock                        25.9   
2              Progressive rock                        22.7   
3     Soundtrack/R&B, soul, pop                        27.4   
4   Hard rock, progressive rock                        20.6   
..                          ...                         ...   
72                    Folk rock                        20.0   
73                          R&B                        20.0   
74                   Soundtrack                        20.0   
75                     Pop, R&B                        20.0   
76                    Pop, Rock                        20.0   

    Claimed sales (millions)  
0                       65.0  
1                       50.0  
2                       45.0  
3                       44.0  
4                       43.0  
..                       ...  
72                       NaN  
73                       NaN  
74                       NaN  
75                       NaN  
76                       NaN  

[77 rows x 6 columns]

Переименуем столбцы для удобства

In [ ]:
hit_albums = hit_albums.rename(columns={'Certified sales (millions)': 'Certified_Sales', 'Claimed sales (millions)': 'Claimed_Sales'})
hit_albums

Рассчитаем средние сертифицированные продажи для всех альбомов.

In [ ]:
print(sum(hit_albums.Certified_Sales) / len(hit_albums.Certified_Sales))

22.40909090909091


* Создайте новый фрейм данных, содержащий только альбомы с сертифицированными продажами (более 20 миллионов человек купили альбом).

In [ ]:
sertific_albums = hit_albums["Certified_Sales"] > 20

sertific_albumss = hit_albums.loc[sertific_albums]

sertific_albumss['Album']

0                                              Thriller
1                                         Back in Black
2                             The Dark Side of the Moon
3                                         The Bodyguard
4                                       Bat Out of Hell
5                       Their Greatest Hits (1971–1975)
7                                               Rumours
8                                          Come On Over
9                                       Led Zeppelin IV
10                                                  Bad
11                                   Jagged Little Pill
12                                     Falling into You
14                                     Hotel California
19                                                    1
20                                                   21
29                                         Supernatural
30                             Appetite for Destruction
31                                  Gold: Greate


 * Создайте новый фрейм данных на основе набора данных hit_albums, который содержит только исполнителей, у которых в списке есть как минимум два альбома.

Сначала выведем строки с дубликатами

In [ ]:
duplicateRows = hit_albums[hit_albums.duplicated(['Artist'])]
duplicateRows

Artist                                              Album  \
10  Michael Jackson                                                Bad   
14           Eagles                                   Hotel California   
16  Michael Jackson                                          Dangerous   
18      Celine Dion                              Let's Talk About Love   
19      The Beatles                                                  1   
21      The Beatles                                         Abbey Road   
28       Pink Floyd                                           The Wall   
34  Various artists  Grease: The Original Soundtrack from the Motio...   
39          Madonna                                      Like a Virgin   
40          Madonna                                          True Blue   
41     Mariah Carey                                           Daydream   
49  Backstreet Boys                                         Millennium   
54         Bon Jovi                                         Cross Road   
56            Adele                                                 25   
59   Britney Spears                            Oops!... I Did It Again   
65  Michael Jackson          HIStory: Past, Present and Future, Book I   
66  Michael Jackson                                       Off the Wall   
68     Shania Twain                                    The Woman in Me   
69     Shania Twain                                                Up!   
74  Various artists  Flashdance: Original Soundtrack from the Motio...   
75  Whitney Houston                                            Whitney   

    Released                       Genre  Certified_Sales  Claimed_Sales  
10      1987             Pop, funk, rock             20.3           34.0  
14      1976  Rock, soft rock, folk rock             21.5           32.0  
16      1991             Rock, Funk, Pop             17.6           32.0  
18      1997              Pop, Soft rock             19.3           31.0  
19      2000                        Rock             21.6           31.0  
21      1969                        Rock             14.4           30.0  
28      1979            Progressive rock             17.6           30.0  
34      1978                  Soundtrack             28.0            NaN  
39      1984                  Pop, dance             25.0            NaN  
40      1986                         Pop             25.0            NaN  
41      1995                    Pop, R&B             25.0            NaN  
49      1999                         Pop             24.0            NaN  
54      1994                   Hard rock             21.0            NaN  
56      2015              Soul, pop, R&B             20.0            NaN  
59      2000                         Pop             20.0            NaN  
65      1995              Pop, rock, R&B             20.0            NaN  
66      1979            Soul, disco, R&B             20.0            NaN  
68      1995                Country, pop             20.0            NaN  
69      2002   Country, pop, world music             20.0            NaN  
74      1983                  Soundtrack             20.0            NaN  
75      1987                    Pop, R&B             20.0            NaN

Затем из этого датафрейма удалим дубликаты

In [ ]:
df_unique = duplicateRows.drop_duplicates("Artist")
df_unique

Artist                                              Album  \
10  Michael Jackson                                                Bad   
14           Eagles                                   Hotel California   
18      Celine Dion                              Let's Talk About Love   
19      The Beatles                                                  1   
28       Pink Floyd                                           The Wall   
34  Various artists  Grease: The Original Soundtrack from the Motio...   
39          Madonna                                      Like a Virgin   
41     Mariah Carey                                           Daydream   
49  Backstreet Boys                                         Millennium   
54         Bon Jovi                                         Cross Road   
56            Adele                                                 25   
59   Britney Spears                            Oops!... I Did It Again   
68     Shania Twain                                    The Woman in Me   
75  Whitney Houston                                            Whitney   

    Released                       Genre  Certified_Sales  Claimed_Sales  
10      1987             Pop, funk, rock             20.3           34.0  
14      1976  Rock, soft rock, folk rock             21.5           32.0  
18      1997              Pop, Soft rock             19.3           31.0  
19      2000                        Rock             21.6           31.0  
28      1979            Progressive rock             17.6           30.0  
34      1978                  Soundtrack             28.0            NaN  
39      1984                  Pop, dance             25.0            NaN  
41      1995                    Pop, R&B             25.0            NaN  
49      1999                         Pop             24.0            NaN  
54      1994                   Hard rock             21.0            NaN  
56      2015              Soul, pop, R&B             20.0            NaN  
59      2000                         Pop             20.0            NaN  
68      1995                Country, pop             20.0            NaN  
75      1987                    Pop, R&B             20.0            NaN

Только артисты

In [ ]:
print(df_unique["Artist"])

10    Michael Jackson
14             Eagles
18        Celine Dion
19        The Beatles
28         Pink Floyd
34    Various artists
39            Madonna
41       Mariah Carey
49    Backstreet Boys
54           Bon Jovi
56              Adele
59     Britney Spears
68       Shania Twain
75    Whitney Houston
Name: Artist, dtype: object


* Создайте новый фрейм данных, содержащий общую сумму всех сертифицированных продаж за каждый год.

In [ ]:
total_sales = hit_albums.groupby('Released')['Certified_Sales'].sum()
total_sales

Released
1967     13.1
1969     14.4
1970     25.0
1971     54.0
1973     22.7
1976     73.7
1977     67.5
1978     28.0
1979     57.6
1980     25.9
1981     25.0
1982     45.4
1983     62.0
1984    109.6
1985     67.7
1986     53.0
1987    144.5
1988     20.0
1990     19.4
1991     54.2
1992     76.4
1993     42.0
1994     64.0
1995    111.8
1996     48.2
1997     91.0
1999     72.5
2000     89.6
2001     20.0
2002     65.0
2004     20.0
2011     22.3
2015     20.0
Name: Certified_Sales, dtype: float64

Вывод: В 1987 году было больше всего продаж, а именно 144.5. В 1967 наоборот: продаж альбомов было меньше всего, так как тогда музыка была не сильно популярна